src/hard_sphere/mod.rs

```diff
@@ -119,16 +119,47 @@ impl<P> HardSphere<P> {
 }
 
 impl<P: HardSphereProperties> HardSphere<P> {
+    /// Returns the Helmholtz energy, packing fractions, and temperature dependent diameters without redundant calculations.
     #[inline]
-    pub fn helmholtz_energy<D: DualNum<f64> + Copy>(&self, state: &StateHD<D>) -> D {
+    pub fn helmholtz_energy_and_properties<D: DualNum<f64> + Copy>(
+        &self,
+        state: &StateHD<D>,
+    ) -> (D, [D; 4], Array1<D>) {
         let p = &self.parameters;
-        let zeta = p.zeta(state.temperature, &state.partial_density, [0, 1, 2, 3]);
+        let diameter = p.hs_diameter(state.temperature);
+
+        let component_index = p.component_index();
+        let geometry_coefficients = p.geometry_coefficients(state.temperature);
+        let mut zeta = [D::zero(); 4];
+        for i in 0..diameter.len() {
+            for (z, &k) in zeta.iter_mut().zip([0, 1, 2, 3].iter()) {
+                *z += state.molefracs[component_index[i]]
+                    * diameter[i].powi(k)
+                    * (geometry_coefficients[k as usize][i] * FRAC_PI_6);
+            }
+        }
+        let zeta_23 = zeta[2] / zeta[3];
```

Comment:
With this change we can actually just get rid of `HardSphereProperties::zeta_23`, it is not used anywhere else.